### ISSUE TITLE
Following pipx completions guidelines for Mac ends with `command not found: compdef` failure

### ISSUE THREAD

**Issue by m-aciek** [pipxproject/pipx#621]
**Describe the bug**
Following pipx completions guidelines for Mac ends with `command not found: compdef` failure.

**How to reproduce**

    % pipx completions

Follow the guidelines. Namely:

    % autoload -U bashcompinit
    % bashcompinit
    % eval "$(register-python-argcomplete pipx)"
    complete:13: command not found: compdef

**Expected behavior**
Successfully registered autocomplete on Mac's Zsh.

**Comment by itsayellow** [comment]
Which version of macOS are you using?  I'm on macOS 11.2 Big Sur with zsh and I have `compdef` present in my shell.

**Comment by itsayellow** [comment]
This might help:
https://unix.stackexchange.com/a/477909
Maybe you need to put the `autoload` command in your `~/.zshrc` and restart your shell?

**Comment by m-aciek** [comment]
Yes, this helps, thank you. Would it be possible to add to completion guidelines that on Mac one should use compinit instead of `autoload -U bashcompinit && bashcompinit`?

It may be out of scope, but compinit prompts me at every new shell:

    zsh compinit: insecure directories, run compaudit for list.
    Ignore insecure directories and continue [y] or abort compinit [n]?

**Comment by itsayellow** [comment]
Shell completions are very shell dependent and not very user-friendly, so our completion instructions just mirror what the python `argcomplete` package documentation suggests.  But it is worth looking into.

The `insecure directories` can be remedied by running `compaudit` once and doing what it suggests.  If I remember correctly it just involves changing the permissions of some directories in ways that are harmless to you.

**Comment by m-aciek** [comment]
Thank you for the support. For what it's worth removing [write access for group](https://stackoverflow.com/a/22753363/1161025) on zsh files solved my problem.

**Comment by m-aciek** [comment]
Actually, after changing `bashcompinit` to `compinit` I got:

    % source ~/.zshrc 
    (eval):46: command not found: complete

Putting [both `bashcompinit` and `compinit`](https://github.com/eddiezane/lunchy/issues/57#issuecomment-448588918) into `.zshrc` finally solved it.

**Comment by gordonje** [comment]
I'm on macOS 11.2.1. I only needed `compinit` as advised for [homebrew's zsh completion](https://docs.brew.sh/Shell-Completion#configuring-completions-in-zsh).

Here are the relevant lines from my ~/.zshrc:

```sh
if type brew &>/dev/null; then
  FPATH=$(brew --prefix)/share/zsh/site-functions:$FPATH

  autoload -Uz compinit
  compinit
fi

eval "$(register-python-argcomplete pipx)"
```

**Comment by baggiponte** [comment]
> I'm on macOS 11.2.1. I only needed compinit as advised for homebrew's zsh completion.

I am on macOS 11.5 and, unfortunately, this was not enough of its own.

Adding `autoload -U bashcompinit && bashcompinit` and then `eval "$(register-python-argcomplete pipx)"` at the end of my `.zshrc` worked. Then I modified my `.zshrc` file to include other things (pure prompt and more zstyles) and now it does not work anymore (the file is [here](https://github.com/baggiponte/dotfiles/blob/main/zsh/.zshrc)).

The weird thing is that if I run again `eval "$(register-python-argcomplete pipx)"` after the shell startup, autocompletion becomes available. I cannot really explain why this happens - after all, this line already comes after `bashcompinit`. I don't know if I should simply put the `eval` statement in `.zprofile`.

Any clue? Does `bashcompinit` need to go before or after `compinit`? I thought it was meant to come afterwards, as noted [here](https://stackoverflow.com/a/67161186/12445701).

**Comment by baggiponte** [comment]
Hi, I finally dug deeper into plugin manager and completion stuff and finally solved my problem up here. TLDR: completion works with the following, i.e. when `bashcompinit` is autoloaded and run *before* evaluating the argcomplete for `pipx`:

```bash
# autoload -Uz compinit && compinit             # not needed as we are using zim
autoload -Uz bashcompinit && bashcompinit
command -v pipx > /dev/null && eval "$(register-python-argcomplete pipx)"
```

`autoload -Uz compinit && compinit` is commented out as I am using the zim package manager which already runs it.

**Comment by cs01** [comment]
Nice, glad you got it working. If there are any changes or doc updates you think would be good, feel free to make a PR.

**Comment by itskando** [comment]
From @philgyford :
```
autoload -U +X bashcompinit && bashcompinit
autoload -U +X compinit && compinit
eval "$(register-python-argcomplete pipx)"
```
worked for me.
